src/main/java/org/truffleruby/parser/BodyTranslator.java

```diff
@@ -3120,23 +3119,32 @@ private static ArgumentsDescriptor getKeywordArgumentsDescriptor(RubyLanguage la
             return EmptyArgumentsDescriptor.INSTANCE;
         }
 
+        final List<String> keywords = new ArrayList<>();
+        boolean splat = false;
+        boolean nonKeywordKeys = false;
+
         for (ParseNodeTuple pair : keywordHashArgumentNode.getPairs()) {
             final ParseNode key = pair.getKey();
             final ParseNode value = pair.getValue();
 
             if (key instanceof SymbolParseNode &&
                     ((SymbolParseNode) key).getName() != null) {
-                return KeywordArgumentsDescriptor.INSTANCE;
+                keywords.add(((SymbolParseNode) key).getName());
             } else if (key == null && value != null) {
                 // A splat keyword hash
-                return KeywordArgumentsDescriptor.INSTANCE;
+                splat = true;
             } else {
                 // For non-symbol keys
-                return KeywordArgumentsDescriptor.INSTANCE;
+                nonKeywordKeys = true;
             }
         }
 
-        return EmptyArgumentsDescriptor.INSTANCE;
+        if (splat || nonKeywordKeys || !keywords.isEmpty()) {
+            return language.keywordArgumentsDescriptorManager
+                    .getArgumentsDescriptor(keywords.toArray(n -> new String[n]));
```

Comment:
`.toArray(EMPTY_STRING_ARRAY)` is better, because it will avoid allocating a new array if keywords is empty.